### PR TITLE
Allow GPIO example built with Linux v6.10+

### PR DESCRIPTION
**Pull request by jeremy90307**
gpio_request_array() and gpio_free_array() were
removed in Kernel v6.10-rc1. Use gpio_request()
and gpio_free() instead to make the example run
successfully on v6.10+.

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
Use preprocessor directives for the compatibility among several Linux kernel versions.

**Comment by jeremy90307** [comment]
I understand. I will fix it by using preprocessor directives.

**Comment by jserv** [comment]
> I understand. I will fix it by using preprocessor directives.

Ensure that LKM can be built with Linux v5.15, v6.1, and v6.6+.

File: examples/bh_threaded.c [hunk 1]
```diff
@@ -13,6 +13,14 @@
 #include <linux/gpio.h>
 #include <linux/delay.h>
 #include <linux/interrupt.h>
+#include <linux/version.h>
+
+/* gpio_request_array() and gpio_free_array() 
+ * were removed in kernel v6.10-rc1 .
+ */
```
Comment (jserv) [inline examples/bh_threaded.c:20]: Mention in git commit messages rather than program comments.

File: examples/bh_threaded.c [hunk 2]
```diff
@@ -13,6 +13,14 @@
...
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 10, 0)
+#define NO_ARRAY
```
Comment (jserv) [inline examples/bh_threaded.c:22]: The macro identifier is misleading. Refine its name.

File: examples/bh_threaded.c (hunk 1 above)
Comment (jserv) [reply examples/bh_threaded.c:20]: You should denote the commit hash associated with the removal of `gpio_request_array` and `gpio_free_array`.

**Review by jserv — CHANGES_REQUESTED** [review]
Fix conflicts and use `git rebase -i` to squash commits.

**Review by jserv — CHANGES_REQUESTED** [review]
Read https://cbea.ms/git-commit/ carefully and enforce the rules.

**Comment by jeremy90307** [comment]
Thank you, I will make the necessary corrections.

File: examples/bh_threaded.c (hunk 2 above)
Comment (jeremy90307) [reply examples/bh_threaded.c]: I have replaced `NO_ARRAY` with `GPIO_NO_REQUEST_ARRAY` to make it clearer.

File: examples/bh_threaded.c (hunk 1 above)
Comment (jserv) [reply examples/bh_threaded.c]: No, always squash and refine the messages.

Comment (jserv) [reply examples/bh_threaded.c]: Read https://cbea.ms/git-commit/ carefully and enforce the rules. You should refine the git commit messages.

Comment (jserv) [reply examples/bh_threaded.c]: > I have removed this comment and mentioned it in the commit message.

You must refer to the exact git commit hash and summarize the rationale.

File: examples/bh_threaded.c [hunk 3]
```diff
@@ -13,6 +13,11 @@
 #include <linux/gpio.h>
 #include <linux/delay.h>
 #include <linux/interrupt.h>
+#include <linux/version.h>
+
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 10, 0)
+#define GPIO_NO_REQUEST_ARRAY
```
Comment (jserv) [inline examples/bh_threaded.c:19]: Rename to `NO_GPIO_REQUEST_ARRAY`.

File: examples/bh_threaded.c [hunk 4]
```diff
@@ -50,29 +55,49 @@ static int __init bottomhalf_init(void)
...
+        pr_err("Unable to request GPIOs for BUTTONs: %d\n", ret);
+        goto Botton1Error;
+    }
+#else
     ret = gpio_request_array(buttons, ARRAY_SIZE(buttons));
 
     if (ret) {
         pr_err("Unable to request GPIOs for BUTTONs: %d\n", ret);
-        goto fail1;
+        goto BottonError;
```
Comment (jserv) [inline examples/bh_threaded.c:90]: Check the file `CONTRIBUTING.md` for consistent style.

File: examples/bh_threaded.c [hunk 5]
```diff
@@ -50,29 +55,49 @@ static int __init bottomhalf_init(void)
...
+#endif
 
     pr_info("Current button1 value: %d\n", gpio_get_value(buttons[0].gpio));
 
     ret = gpio_to_irq(buttons[0].gpio);
 
     if (ret < 0) {
         pr_err("Unable to request IRQ: %d\n", ret);
-        goto fail2;
+        goto IRQ0Error;
```
Comment (jserv) [inline examples/bh_threaded.c:100]: `IRQ0Error` was not better.

Comment (jeremy90307) [reply examples/bh_threaded.c]: I originally thought fail1, fail2, and fail3 would reduce the program's readability, but it has been fixed back.

File: examples/bh_threaded.c (hunk 3 above)
Comment (jeremy90307) [reply examples/bh_threaded.c]: I have fixed it.

**Comment by jeremy90307** [comment]
Is there anything to improve in this commit? Thank you.

**Comment by jserv** [comment]
> Is there anything to improve in this commit?

Carefully study the principles of https://cbea.ms/git-commit/ . When writing a commit message, use the imperative mood in the subject line, specifically employing precise action verbs. For instance, use "Allow" rather than passive constructions.
Ensure your commit message body is wrapped at 72 characters without cutting off words mid-line. More critically, focus on explaining the what and why of your changes, not the how. The message should provide substantive context about the rationale behind the code modification.

Take commit d2d54ca63944675a6ee17747eda5947207001426 for example. We prefer the notation `commit b8780c363d80 ("sched: remove sleep_on() and friends")`. This format is preferred because it concisely combines the commit hash with a descriptive subject line. The message should go beyond mere description and dive into the underlying reasoning for the change.

When contributing to a technical project, communication should be direct and purposeful. The phrase "Thank you" becomes unnecessary when the primary goal is collaborative technical work. In a professional development context, your interactions should focus on the technical details, code quality, and project improvement.

File: examples/bh_threaded.c (hunk 3 above)
Comment (jserv) [reply examples/bh_threaded.c]: You don't have to put the message "I have fixed it." Instead, simply click "Resolve conversation" via GitHub web interface once you complete.

**Review by jserv — CHANGES_REQUESTED** [review]
Compose Git commit messages in the third person, and include a test summary specific to Raspberry Pi deployment or testing.

**Comment by jserv** [comment]
Refine the git commit message as following:
> Commit dbcedec ("gpiolib: legacy: Remove unused gpio_request_array()  
> and gpio_free_array()") indicated that there are no more users of these  
> functions; therefore, they were removed.  

And address the considerations of removal fro Linux kernel developers' perspectives.

**Comment by jeremy90307** [comment]
https://github.com/user-attachments/assets/bc43c25b-f02a-43c7-9bd1-a178f94b99f5
Successfully ran on Linux v6.12.1

**Comment by jeremy90307** [comment]
> > https://github.com/user-attachments/assets/bc43c25b-f02a-43c7-9bd1-a178f94b99f5 Successfully ran on Linux v6.12.1.
> 
> You should explain the testing procedure within git commit messages as well.

Should I explain the setup for the GPIO connections?

**Comment by jeremy90307** [comment]
My testing steps only include compiling, connecting an LED and a button, and finally loading the module.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `master` branch to pass CI pipeline.

**Review by jserv — CHANGES_REQUESTED** [review]
Replace `Resolves:#285` with shorter `Close #285`.

**Comment by jserv** [comment]
Thank @jeremy90307 for contributing!

**Pull request merged**